cpp/src/decisiontree/batched-levelalgo/builder.cuh

```diff
@@ -443,29 +444,33 @@ struct Builder {
   }
 
   // Set the leaf value predictions in batch
-  void SetLeafPredictions(std::vector<NodeT>* tree,
+  void SetLeafPredictions(std::shared_ptr<DT::TreeMetaDataNode<DataT, LabelT>> tree,
                           const std::vector<InstanceRange>& instance_ranges)
   {
-    // do this in batch to reduce peak memory usage in extreme cases
-    std::size_t max_batch_size = min(std::size_t(100000), tree->size());
-    rmm::device_uvector<NodeT> d_tree(max_batch_size, handle.get_stream());
-    rmm::device_uvector<InstanceRange> d_instance_ranges(max_batch_size, handle.get_stream());
+    tree->vector_leaf.resize(tree->sparsetree.size() * input.numOutputs);
+    ASSERT(tree->sparsetree.size() == instance_ranges.size(),
+           "Expected instance range for each node");
+    rmm::device_uvector<NodeT> d_nodes(tree->sparsetree.size(), handle.get_stream());
+    rmm::device_uvector<DataT> d_leaves(tree->vector_leaf.size(), handle.get_stream());
+    rmm::device_uvector<InstanceRange> d_instance_ranges(tree->sparsetree.size(),
+                                                         handle.get_stream());
     ObjectiveT objective(input.numOutputs, params.min_impurity_decrease, params.min_samples_leaf);
-    for (std::size_t batch_begin = 0; batch_begin < tree->size(); batch_begin += max_batch_size) {
-      std::size_t batch_end  = min(batch_begin + max_batch_size, tree->size());
-      std::size_t batch_size = batch_end - batch_begin;
-      raft::update_device(
-        d_tree.data(), tree->data() + batch_begin, batch_size, handle.get_stream());
-      raft::update_device(d_instance_ranges.data(),
-                          instance_ranges.data() + batch_begin,
-                          batch_size,
-                          handle.get_stream());
-      size_t smemSize = sizeof(BinT) * input.numOutputs;
-      int num_blocks  = batch_size;
-      leafKernel<<<num_blocks, TPB_DEFAULT, smemSize, handle.get_stream()>>>(
-        objective, input, d_tree.data(), d_instance_ranges.data());
-      raft::update_host(tree->data() + batch_begin, d_tree.data(), batch_size, handle.get_stream());
-    }
+    raft::update_device(
+      d_nodes.data(), tree->sparsetree.data(), tree->sparsetree.size(), handle.get_stream());
+    raft::update_device(d_instance_ranges.data(),
+                        instance_ranges.data(),
+                        tree->sparsetree.size(),
+                        handle.get_stream());
+    raft::update_device(
```

Comment:
Is this copy operation needed? Isn't `d_leaves` is the output variable for the kernel?

Reply:
Changed to a cudaMemsetasync just so we don't get any garbage values in the unused part of the vector.